### ISSUE TITLE
Support for Redirect IPv6 extended communities (for Flowspec Redirect-IPv6)

### ISSUE THREAD

**Issue by pguibert6WIND** [Exa-Networks/exabgp#927]
Hello,

Is the RFC 5701 supported ?
If not, this would be good to have in order to have support for Redirect-IPv6 from [0]

Thanks,
Philippe
[0] https://tools.ietf.org/html/draft-ietf-idr-flow-spec-v6-09

**Comment by pguibert6WIND** [comment]
which draft overrides the other one : draft-ietf-idr-flow-spec-v6-09 or draft-ietf-idr-flowspec-redirect-ip ?

**Comment by thomas-mangin** [comment]
I do not believe that https://tools.ietf.org/html/rfc5701 was implemented. What you are finding is the Flowspec next hop redirect RFC. It should however be quite straight forward to implement.

**Comment by thomas-mangin** [comment]
I have not implemented redirect-to-nexthop-ietf - as far as the author @elindsey tested it is working - https://github.com/Exa-Networks/exabgp/pull/923

**Comment by elindsey** [comment]
> I can however see a 20 byte ipv6 ext community forged, but with wrong type and subtype.

It is the correct type and subtype, but you're looking for a different feature. 🙂 

The patchset recently added is for redirect to next hop (draft-ietf-idr-flowspec-redirect-ip), this is for rewriting a next hop. This is action type 0x000c, as documented here https://www.iana.org/assignments/bgp-extended-communities/bgp-extended-communities.xhtml (search for 0x000c).

What you're asking about is redirect ipv6, action type 0x800b. This is for redirecting traffic into a VRF, picked by a specific ipv6 address being in its import policy. This is not implemented.

In general, rfcs that say nexthop are going to be a simple destination IP rewrite. Anything that says redirect without "nexthop" in the name is going to be some form of local VRF redirection. Many of those get complicated and have mixed degrees of support.

**Comment by elindsey** [comment]
As for rfc5701 specifically - support for it was added to be able to send the ipv6 extended community for the 0x000c next hop action. I don't believe anything else in the exa codebase currently uses it, but it would be easy to modify if you want to support additional ipv6 extended communities.

**Comment by pguibert6WIND** [comment]
I would be ok to specify the command.
what about the following :
`redirect-to-nexthop-vrf-rt-ipv6 <IPV6>:<AS2B>`

for the implementation, I also would be happy to help for that; alittle guidance could be helpful however.
thanks,

**Comment by elindsey** [comment]
I don't speak for exa or Thomas, so anything he says overrides me - but I'll voice my thoughts

> what about the following : redirect-to-nexthop-vrf-rt-ipv6 <IPV6>:<AS2B>

I would first see if this functionality can be attached to the existing `redirect` config directive, which handles the original rfc 5575 0x8008 action. If possible I would extend that same config directive to now handle the ipv6:asn notation.

`redirect-to-nexthop`/`redirect-to-nexthop-ietf` exists because of the unfortunate competing drafts and to disambiguate from the vrf forms. That doesn't apply here, so hopefully you won't need to go down the route of adding a new config directive.

[This recent pull request is a useful reference.](https://github.com/Exa-Networks/exabgp/pull/923)

[This is where the `redirect` directive is parsed, and where I would look to see if you can add ipv6:asn parsing.](https://github.com/elindsey/exabgp/blob/1136ebeedb7d583ea020090c028ceec5257a6426/lib/exabgp/configuration/flow/parser.py#L276) 

[This commented out stub was going to add support for an ipv4:asn vrf redirect and may be relevant.](https://github.com/elindsey/exabgp/blob/1136ebeedb7d583ea020090c028ceec5257a6426/lib/exabgp/bgp/message/update/attribute/community/extended/traffic.py#L284)

**Comment by pguibert6WIND** [comment]
I have to admit that the format is equal to me, provided that the feature can be put into exabgp.
so your counter proposal would suit me.
thanks.

**Comment by thomas-mangin** [comment]
@elindsey thank you for your help figuring out what the type and sub-type needed to be :p
I noticed that there was a bug with you code (as the registration of the two differents Attributes (normal and the v6 you added) was done within the same namespace. I fixed it with the patch above moving registered_extended into the sub-classes to give them both distinct registration.

@pguibert6WIND can you please try and tell us if it is working as you expect. I have no way to test. If it is ok, then I will add some code for self-testing the generation and decoding of the attributes. The code is untested though.

**Comment by elindsey** [comment]
Ahh, thanks for the fix. 🙂 

I think the 0x0002 in the patch is incorrect. RFC 5701 specifies a route target format of type 0x0002, but draft-ietf-idr-flow-spec-v6-08 is the actual flowspec redirect ipv6 action that happens to reuse that format for a different purpose with type 0x800B. 0x80 is from the experimental use range, which is why I assume it's not in the IANA registry - this redirect ipv6 draft doesn't look finished or adopted.

@pguibert6WIND have you seen any vendor gear that claims support for draft-ietf-idr-flow-spec-v6-08? I haven't been able to turn up any.

**Comment by thomas-mangin** [comment]
@elindsey happy to change the 0x0002 to whatever is seen in production ...

**Comment by thomas-mangin** [comment]
@elindsey perhaps I am mixing redirect with adding an IPv6 RT and Origin support ..It has been quite a few years since I last looked at the community RFCs.

**Comment by pguibert6WIND** [comment]
Hi @thomas-mangin ,
so previously, I made a patch to replace the route ttarget redirect IP with route target redirect VRF like that one:
```
 class TrafficNextHopIPv6IETF (ExtendedCommunityIPv6):
-       COMMUNITY_TYPE = 0x00
-       COMMUNITY_SUBTYPE = 0x0C
+       COMMUNITY_TYPE = 0x80
+       COMMUNITY_SUBTYPE = 0x0B
```
the behaviour is ok from my side.
now with the patch you submitted, I made change to the exabgp config:
```
-redirect-to-nexthop-ietf 33::2;
+redirect 33::2;
```
and I tried to run exabgp, but with no success.
```
18:38:34 | 14264  | configuration | not reloaded, no change found in the configuration
18:38:34 | 14264  | configuration | problem parsing configuration file line 15
18:38:34 | 14264  | configuration | error message: illegal IP address string passed to inet_pton
```
did I make someting wrong?

**Comment by pguibert6WIND** [comment]
adding to this, currently it is possible to have the following as for simpson draft:
```
redirect IPV6;
```
My expectation is that redirect either takes an IP (v4 or v6 like for simpson draft with IP included in NLRI)  or an IPv6 route target ( the new draft for redirect vrf).

**Comment by thomas-mangin** [comment]
Need to look how I can do the parsing ..

**Comment by thomas-mangin** [comment]
`redirect IPV6:asn` should now work as you want

**Comment by thomas-mangin** [comment]
not sure what I should do to close this ...

**Comment by pguibert6WIND** [comment]
Hi Thomas,
please give me a bit more time to clarify better the need.
